File: version4j/src/main/java/com/only4/algorithm/version4j/leetcode/InorderTraversal.java
package com.only4.algorithm.version4j.leetcode;

import com.only4.algorithm.version4j.extra.TreeNode;

import java.util.ArrayList;
import java.util.List;

public class InorderTraversal {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        inorderHelper(root, result);
        return result;
    }

    private void inorderHelper(TreeNode node, List<Integer> result) {
        if (node == null) return;

        // 递归遍历左子树
        inorderHelper(node.left, result);
        // 访问当前节点
        result.add(node.val);
        // 递归遍历右子树
        inorderHelper(node.right, result);
    }
}

File: version4k/src/main/kotlin/com/only4/algorithm/version4k/leetcode/inorderTraversal.kt
package com.only4.algorithm.version4k.leetcode

import com.only4.algorithm.version4k.extra.TreeNode

fun inorderTraversal(root: TreeNode?): List<Int> {
    val result = mutableListOf<Int>()

    fun inorderHelper(node: TreeNode?) {
        if (node == null) return

        // 递归遍历左子树
        inorderHelper(node.left)
        // 访问当前节点
        result.add(node.`val`)
        // 递归遍历右子树
        inorderHelper(node.right)
    }

    inorderHelper(root)
    return result
}
